lib/testing.ex

```diff
@@ -0,0 +1,206 @@
+
+  # Asserting that no file was deleted
+  refute_deleted_files()
+  ```
+
+  If you need a more low level API, you can directly use the `Uploadex.TestStorage` functions.
+  """
+
+  import ExUnit.Assertions, only: [assert: 2]
+
+  alias Uploadex.TestStorage
+
+  @doc """
+  Imports all functions from `Uploadex.Testing`
+  and defines a setup callback to start the `Uploadex.TestStorage`.
+  """
+  defmacro __using__(_opts) do
+    quote do
+      import Uploadex.Testing
+
+      setup :start_test_storage
+    end
+  end
+
+  @doc """
+  Starts the `Uploadex.TestStorage`.
+
+  You can use it as a `ExUnit` setup callback:
+
+      setup :start_test_storage
+
+  """
+  @spec start_test_storage(context :: map()) :: :ok
+  def start_test_storage(_ctx \\ %{}) do
+    TestStorage.start_link()
+    :ok
+  end
+
+  @doc """
+  Asserts that the given `expected_count` number of files was stored.
+  """
+  @spec assert_stored_files_count(expected_count :: integer) :: true
+  def assert_stored_files_count(expected_count) when is_integer(expected_count) do
+    stored_files = TestStorage.get_stored()
+    stored_count = Enum.count(stored_files)
+
+    error_message = """
+    Expected #{expected_count} files stored.
+
+    Got #{stored_count} instead:
+
+    #{inspect(stored_files)}
+    """
+
+    assert expected_count == stored_count, error_message
+  end
+
+  @doc """
+  Asserts that the given `expected_count` number of files was deleted.
+  """
+  @spec assert_deleted_files_count(expected_count :: integer) :: true
+  def assert_deleted_files_count(expected_count) when is_integer(expected_count) do
+    deleted_files = TestStorage.get_deleted()
+    deleted_count = Enum.count(deleted_files)
+
+    error_message = """
+    Expected #{expected_count} files deleted.
+
+    Got #{deleted_count} instead:
+
+    #{inspect(deleted_files)}
+    """
+
+    assert expected_count == deleted_count, error_message
+  end
+
+  @doc """
+  Asserts that no file was stored.
+  """
+  @spec refute_stored_files() :: true
+  def refute_stored_files() do
+    stored_files = TestStorage.get_stored()
+
+    error_message = """
+    Expected no files stored.
+
+    Got: #{inspect(stored_files)}
+    """
+
+    assert [] == stored_files, error_message
+  end
+
+  @doc """
+  Asserts that no file was deleted.
+  """
+  @spec refute_deleted_files() :: true
+  def refute_deleted_files() do
+    deleted_files = TestStorage.get_deleted()
+
+    error_message = """
+    Expected no files deleted.
+
+    Got: #{inspect(deleted_files)}
+    """
+
+    assert [] == deleted_files, error_message
+  end
+
+  @doc """
+  Asserts that all the files in `expected_files` were stored
+  in the exact order that they are passed in the list.
+
+  ## Options
+
+    * `ignoring_order`: if `true`, ignores the order of the files when asserting (defaults to `false`)
+
+  """
+  @spec assert_stored_files(expected_files :: list(), opts :: [ignoring_order: boolean()]) :: true
```

Comment:
with this signature, wouldn't dialyzer always expect `ignoring_order` to be passed?

Reply:
No, I checked it :)